File: 백준/Gold/2110. 공유기 설치/공유기 설치.py
import bisect
from typing import List


class Solution:
    def install_router(self, lst: List[int], C: int) -> int:

        lst.sort()

        low = 1 # 라우터간 최솟값은 1
        high = lst[-1] - lst[0] # 라우터간 최대값은 양 끝값의 거리
        min_gap = 0

        while low <= high:
            mid = (low + high) // 2

            # 공유기 설치 개수 & 현재 공유기 위치
            cnt = 1
            cur = lst[0]

            for idx in range(1, len(lst)):
                if lst[idx] >= cur + mid:
                    cnt += 1
                    cur = lst[idx]

            # 공유기 거리가 mid 일때 설치된 공유기 수가 n을 만족한다면
            if cnt >= C:
                min_gap = mid
                low = mid + 1
            else:
                high = mid - 1

        return min_gap


if __name__ == '__main__':
    s = Solution()

    N, C = input().split(" ")

    lst = []

    for _ in range(int(N)):
        temp = int(input())
        lst.append(temp)

    print(s.install_router(lst, int(C)))

File: 백준/Gold/2110. 공유기 설치/공유기 설치.java
import java.io.*;
import java.util.Arrays;

import static java.lang.Integer.parseInt;

public class Main{
    public static int install_router(int[] lst, int C) {

        // 배열 정렬
        Arrays.sort(lst);

        int low = 1;
        int high = lst[lst.length-1] - lst[0];
        int minGap = 0;
        while (low <= high) {
            int mid = (low + high) / 2;

            int cur = lst[0];
            int cnt = 1;

            for(int i=1; i<lst.length; i++) {

                if(lst[i] >= cur + mid) {
                    cur = lst[i];
                    cnt++;
                }
            }

            if (cnt >= C) {
                minGap = mid;
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return minGap;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String str = br.readLine();
        int N = parseInt(str.split(" ")[0]);
        int C = parseInt(str.split(" ")[1]);

        int[] lst = new int[N];

        for (int i = 0; i < N; i++) {
            lst[i] = parseInt(br.readLine());
        }

        bw.write(install_router(lst, C)+"");
        bw.flush();
    }
}